src/util/permissions.tsx

```diff
@@ -0,0 +1,311 @@
+      resourceEntitlementOptions[resource] = [defaultOption];
+    }
+
+    resourceEntitlementOptions[resource].push({
+      value: permission.entitlement,
+      label: permission.entitlement,
+    });
+  }
+
+  const resourceOptions: OptionHTMLAttributes<HTMLOptionElement>[] = [
+    defaultOption,
+  ];
+
+  for (const resource in resourceEntitlementOptions) {
+    // split entitlement options into two based on 'can_' prefix
+    const genericEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+    const granularEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+
+    for (const entitlementOption of resourceEntitlementOptions[resource]) {
+      const entitlementValue = entitlementOption.value as string;
+      if (entitlementValue.includes("can_")) {
+        granularEntitlements.push(entitlementOption);
+        continue;
+      }
+
+      genericEntitlements.push(entitlementOption);
+    }
+
+    genericEntitlements.sort(sortOptions);
+    granularEntitlements.sort(sortOptions);
+
+    // add disabled option as a delimiter between the two sets of entitlements
+    if (genericEntitlements.length > 1 && granularEntitlements.length) {
+      granularEntitlements.unshift({
+        disabled: true,
+        label: "Granular entitlements",
+        value: "",
+      });
+    }
+    resourceEntitlementOptions[resource] =
+      genericEntitlements.concat(granularEntitlements);
+
+    resourceOptions.push({
+      value: resource,
+      label: extractResourceNameFromUrl(resourceType, resource),
+    });
+  }
+
+  resourceOptions.sort(sortOptions);
+
+  return { resourceOptions, resourceEntitlementOptions };
+};
+
+// refer to api spec to see how the names can be extracted from resource url
+// https://documentation.ubuntu.com/lxd/en/latest/api/
+export const extractResourceNameFromUrl = (
+  resourceType: string,
+  url: string,
+) => {
+  const urlSegments = url.split("/");
+  let resourceName = "";
+  switch (resourceType) {
+    case "server":
+      resourceName = "server";
+      break;
+    case "storage_volume":
+      resourceName = urlSegments[6].split("?")[0];
+      break;
+    case "identity":
+    case "storage_bucket":
+      resourceName = urlSegments[5].split("?")[0];
+      break;
+    case "group":
+    case "image_alias":
+      resourceName = urlSegments[4].split("?")[0];
+      break;
+    case "network":
+    case "network_acl":
+    case "network_zone":
+    case "instance":
+    case "image":
+    case "profile":
+    case "storage_pool":
+    case "project":
+    case "certificate":
+      resourceName = urlSegments[3].split("?")[0];
+      break;
+    default:
+      resourceName = urlSegments[0];
+      break;
+  }
+
+  return decodeURIComponent(resourceName);
+};
+
+export const getPermissionId = (
+  permission: LxdPermission & { id?: string },
+) => {
+  if (permission.id) {
+    return permission.id;
+  }
+
+  return permission.entity_type + permission.url + permission.entitlement;
+};
```

Comment:
```suggestion
export const getPermissionId = (permission: LxdPermission) => {
  return permission.entity_type + permission.url + permission.entitlement;
};
```
This can be simpler, no?

Reply:
Aaah right, this was because previously I had the resource state storing the extracted name instead of the full url which caused issues, although that's resolved but I forgot to remove this, thanks! :+1: